components/string.rst

```diff
@@ -0,0 +1,378 @@
+.. index::
+   single: String
+   single: Components; String
+
+The String Component
+====================
+
+    The String component provides a single object-oriented API to work with
+    three "unit systems" of strings: bytes, code points and grapheme clusters.
+
+Installation
+------------
+
+.. code-block:: terminal
+
+    $ composer require symfony/string
+
+.. include:: /components/require_autoload.rst.inc
+
+What is a String?
+-----------------
+
+You can skip this section if you already know what a *"code point"* or a
+*"grapheme cluster"* are in the context of handling strings. Otherwise, read
+this section to learn about the terminology used by this component.
+
+Languages like English require a very limited set of characters and symbols to
+display any content. Each string is a series of characters (letters or symbols)
+and they can be encoded even with the most limited standards (e.g. `ASCII`_).
+
+However, other languages require thousands of symbols to display their contents.
+They need complex encoding standards such as `Unicode`_ and concepts like
+"character" no longer make sense. Instead, you have to deal with these terms:
+
+* `Code points`_: they are the atomic unit of information. A string is a series
+  of code points. Each code point is a number whose meaning is given by the
+  `Unicode`_ standard. For example, the English letter ``A`` is the ``U+0041``
+  code point and the Japanese *kana* ``の`` is the ``U+306E`` code point.
+* `Grapheme clusters`_: they are a sequence of one or more code points which are
+  displayed as a single graphical unit. For example, the Spanish letter ``ñ`` is
+  a grapheme cluster that contains two code points: ``U+006E`` = ``n`` (called
+  *"latin small letter N"*) + ``U+0303`` = ``◌̃`` (called *"combining tilde"*).
+* Bytes: they are the actual information stored for the string contents. Each
+  code point can require one or more bytes of storage depending on the standard
+  being used (UTF-8, UTF-16, etc.).
+
+The following image displays the bytes, code points and grapheme clusters for
+the same word written in English (``hello``) and Hindi (``नमस्ते``):
+
+.. image:: /_images/components/string/bytes-points-graphemes.png
```

Comment:
what's the source of the image? I find it difficult to understand (and it contains approximations, eg a single code point above a combined character)
I'd suggest using the same example as above: accents are easy to grasp.
Not that `ñ` is also a single code point so the text might benefit from a "can be represented as two code points"

Reply:
The image is custom made. I don't want to use accents or letter like `ñ` or `ç` precisely because what you said: they can be expressed as a single code point and as a sequence of two code points. However, using a word like `नमस्ते` removes those problems. (By the way, it's the same word used by Rust in their docs to explain all this).

About the "approximations" ... I thought they were correct because of this:

```php
$string = b('नमस्ते');

echo "\n\nBytes: ";
for ($i = 0; $i < $string->length(); $i++) {
    echo $string->byteCode($i)." ";
}

echo "\n\nCode Points: ";
for ($i = 0; $i < $string->toCodePointString()->length(); $i++) {
    echo $string->toCodePointString()->codePoint($i)." ";
}

echo "\n\nGraphemes: ";
for ($i = 0; $i < $string->toUnicodeString()->length(); $i++) {
    echo $string->toUnicodeString()->codePoint($i)." ";
}
```

Output:

```
Bytes: 224 164 168 224 164 174 224 164 184 224 165 141 224 164 164 224 165 135

Code Points: 2344 2350 2360 2381 2340 2375

Graphemes: 2344 2350 2360 2340
```

So, is this correct? Thanks.

Reply:
codePoint returns the code point of the first character
`Graphemes: 2344 2350 2360+2381 2340+2375` would be more accurate then

Reply:
Ah! Perfect. I'll update this. Thank you.